### PR TITLE
Course selection journey for duplicate course selection

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

A candidate can add up to 4 concurrent active applications to their application form per recruitment cycle. In the process of choosing a course choice for an application, it's possible for them to choose a course for which they have already made an application.

If they choose a course for which they have already made an application in the current recruitment cycle, a database constraint will not allow them to add a duplicate course selection.

This PR provides a user flow for candidates attempting to add a duplicate course choice.

## Changes proposed in this pull request

 - Add a new step `DuplicateCourseSelectionStep` to the `CourseSelectionWizard` 
 - Add `CourseChoiceValidator` that the `WhichCourseAreYouApplyingToStep` can use to determine the next step in the wizard.

There are some issues with the back link on the duplicate selection page during an edit. This can be fixed in another PR.

## Screenshot

Screenshot from 2023-09-25 14-34-57

## Guidance to review

I noticed I used `CourseChoice` and `CourseSelection`. Which should be used?

## Link to Trello card

[Trello Ticket](https://trello.com/c/zjjWBIx4/444-ca-course-selection-apply-for-same-course)

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [x] API release notes have been updated if necessary
- [x] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

**Comment by inulty-dfe** [comment]
## Testing Review App

Candidate with carried over application in the 2024 cycle: https://apply-review-8566.test.teacherservices.cloud/support/applications/44

File: app/forms/candidate_interface/continuous_applications/course_choice_validator.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module CandidateInterface
+  module ContinuousApplications
+    class CourseChoiceValidator < ActiveModel::Validator
+      def validate(record)
+        return unless record.wizard.current_application
+        return if record.wizard.edit?
+
+        exists = record.wizard.current_application.application_choices.joins(:course_option).exists?(course_option: { course_id: record.course.id })
```
Comment (vassyz) [inline app/forms/candidate_interface/continuous_applications/course_choice_validator.rb:8]: Would this work?

```ruby
record.wizard.current_application.contains_course?(record.course)
```

File: app/forms/candidate_interface/continuous_applications/which_course_are_you_applying_to_step.rb [hunk 2]
```diff
@@ -5,6 +5,8 @@ class WhichCourseAreYouApplyingToStep < DfE::WizardStep
       attr_accessor :provider_id, :course_id
       validates :provider_id, :course_id, presence: true
 
+      validate CandidateInterface::ContinuousApplications::CourseChoiceValidator.new, on: :course_choice
```
Comment (vassyz) [inline app/forms/candidate_interface/continuous_applications/which_course_are_you_applying_to_step.rb:8]: If you move the `CourseChoiceValidator` to `app/validators` you could replace this with `      validates self, course_choice: true, on: :course_choice`

File: app/forms/candidate_interface/continuous_applications/course_choice_validator.rb (hunk 1 above)
Comment (inulty-dfe) [reply app/forms/candidate_interface/continuous_applications/course_choice_validator.rb:8]: Looks like it does the same thing alright!

This doesn't look like the best implementation of this, does it?
```ruby
  def contains_course?(course)
    potential_course_option_ids = CourseOption.where(course_id: course.id).map(&:id)
    current_course_option_ids = application_choices.map(&:course_option_id)

    potential_course_option_ids.intersect?(current_course_option_ids)
  end
```

File: app/controllers/candidate_interface/continuous_applications/course_choices/duplicate_course_selection_controller.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      class DuplicateCourseSelectionController < BaseController
+        before_action :set_course
+
+      private
+
+        def step_params
+          params[current_step] = {}
+          params[current_step][:provider_id] = params[:provider_id]
+          params[current_step][:course_id] = params[:course_id]
+          params
```
Comment (vassyz) [inline app/controllers/candidate_interface/continuous_applications/course_choices/duplicate_course_selection_controller.rb:13]: Personal style preference:

```ruby
params[current_step] = {
  provider_id: params[:provider_id],
  course_id: params[:course_id],
}
params
```

File: app/forms/candidate_interface/continuous_applications/course_choice_validator.rb (hunk 1 above)
Comment (vassyz) [reply app/forms/candidate_interface/continuous_applications/course_choice_validator.rb:8]: Yes, could probably be improved.

Comment (vassyz) [reply app/forms/candidate_interface/continuous_applications/course_choice_validator.rb:8]: What about: `record.wizard.current_application.courses.exists?(record.course.id)`?

File: app/forms/candidate_interface/continuous_applications/which_course_are_you_applying_to_step.rb [hunk 4]
```diff
@@ -24,25 +26,35 @@ def previous_step
...
           { provider_id:, course_id: }
         elsif multiple_sites?
           { provider_id:, course_id:, study_mode: }
         end
       end
 
       def edit_next_step_path_arguments
-        if completed?
+        if duplicate_course?
+          default_path_arguments.merge({ provider_id:, course_id: })
```
Comment (vassyz) [inline app/forms/candidate_interface/continuous_applications/which_course_are_you_applying_to_step.rb:56]: `default_path_arguments.merge(provider_id:, course_id:)` No need for curly brackets, but does this code ever get called?

**Review by MaeveRoseveare — COMMENTED** [review]
This looks good @inulty-dfe I noticed 2 things (one unrelated to this work actually)

1. I think the H1 is pulling in the Course name, qualification (QTS) and the study mode here? I think it would be clearer to say "You've already applied to [Biology 83SL] at [Provider name]

Screenshot 2023-09-21 at 15 09 01

2. Unrelated but I noticed 2 full stops on both of the blocking messages. Perhaps a separate ticket?

Screenshot 2023-09-21 at 15 10 01

**Comment by vassyz** [comment]
Don't think there is a system spec for changing a course, but not sure it will work because of [this](https://github.com/DFE-Digital/apply-for-teacher-training/pull/8566/files#diff-7cbd1efb8bef7d74ab587f29034ecbe4df3752bdbccb5af9f047650de75a2a03R6) and [on: :course_choice](https://github.com/DFE-Digital/apply-for-teacher-training/pull/8566/files#diff-73bdb2a003d1de1baae38f576c0344097264f303a68d00dc7fa2bf77cbfe3780R8).

**Comment by JR-G** [comment]
> Shouldn't this PR also implement the withdrawn/rejected logic?

Now we've roughly agreed on an approach If we can cover it here then great, but I'm also happy for that part to go into a subsequent PR (which I'll write up a card for) as the priority is getting this merged in. 

The rationale is that with Find opening on the 3rd and then Apply opening on the 10th, I feel there's a higher chance we'll sooner encounter a candidate going down the flow of selecting the a previous chosen choice, than we would of a candidate selecting a course, getting rejected/withdrawing and then wanting to immediately apply.

**Comment by vassyz** [comment]
> > Shouldn't this PR also implement the withdrawn/rejected logic?
> 
> Now we've roughly agreed on an approach If we can cover it here then great, but I'm also happy for that part to go into a subsequent PR (which I'll write up a card for) as the priority is getting this merged in.
> 
> The rationale is that with Find opening on the 3rd and then Apply opening on the 10th, I feel there's a higher chance we'll sooner encounter a candidate going down the flow of selecting the a previous chosen choice, than we would of a candidate selecting a course, getting rejected/withdrawing and then wanting to immediately apply.

Wasn't quire sure what the requirements were, but the Trello ticket has this scenario:

>User adds Biology UCL and it is unsuccessful (rejected / withdraw, declined, etc) -> user tries to add again the same course ->  if they’ve been rejected then they should be allowed to apply again to the same course/provider so they have a chance to improve their application

Is it not just a matter of checking if the state is not one of the `UNSUCCESSFUL_STATES` in the validator?

**Comment by JR-G** [comment]
> > > Shouldn't this PR also implement the withdrawn/rejected logic?
> > 
> > 
> > Now we've roughly agreed on an approach If we can cover it here then great, but I'm also happy for that part to go into a subsequent PR (which I'll write up a card for) as the priority is getting this merged in.
> > The rationale is that with Find opening on the 3rd and then Apply opening on the 10th, I feel there's a higher chance we'll sooner encounter a candidate going down the flow of selecting the a previous chosen choice, than we would of a candidate selecting a course, getting rejected/withdrawing and then wanting to immediately apply.
> 
> Wasn't quire sure what the requirements were, but the Trello ticket has this scenario:
> 
> > User adds Biology UCL and it is unsuccessful (rejected / withdraw, declined, etc) -> user tries to add again the same course ->  if they’ve been rejected then they should be allowed to apply again to the same course/provider so they have a chance to improve their application
> 
> Is it not just a matter of checking if the state is not one of the `UNSUCCESSFUL_STATES` in the validator?

That's right! – but we'll also need to remove the `t.index ["application_form_id", "course_option_id"], name: "index_course_option_to_application_form_id", unique: true
` for this to work?

**Comment by vassyz** [comment]
> That's right! – but we'll also need to remove the `t.index ["application_form_id", "course_option_id"], name: "index_course_option_to_application_form_id", unique: true ` for this to work?

Ok, forgot about that. Whatever approach you find more convenient then.

File: app/forms/candidate_interface/continuous_applications/which_course_are_you_applying_to_step.rb (hunk 2 above)
Comment (inulty-dfe) [reply app/forms/candidate_interface/continuous_applications/which_course_are_you_applying_to_step.rb:8]: Move it to app/validators

**Comment by inulty-dfe** [comment]
@MaeveRoseveare 

**"You've already applied" v "You already have an application"**

In reply to your suggestion: 

> I think the H1 is pulling in the Course name, qualification (QTS) and the study mode here? I think it would be clearer to say "You've already applied to [Biology 83SL] at [Provider name]

The current text works for unsubmitted applications. The action of "applying" suggests the application was submitted. This page shows when the candidate chooses a course for which they already have an unsubmitted application as well as submitted.

**Comment by MaeveRoseveare** [comment]
> @MaeveRoseveare
> 
> **"You've already applied" v "You already have an application"**
> 
> In reply to your suggestion:
> 
> > I think the H1 is pulling in the Course name, qualification (QTS) and the study mode here? I think it would be clearer to say "You've already applied to [Biology 83SL] at [Provider name]
> 
> The current text works for unsubmitted applications. The action of "applying" suggests the application was submitted. This page shows when the candidate chooses a course for which they already have an unsubmitted application as well as submitted.

Oops, sorry!!!!! Yes you're right. It should be 'You already have an application...." - but we should still not show the QTS or full time and rather just show the course and provider name. So:

"You already have an application for [Biology XYZ] at [West Midlands Consortium]"

**Review by MaeveRoseveare — APPROVED** [review]
Looks good! I just had that question about if we allow them to add the same application if they've withdrawn the same one already. At the moment you can't.

The ticket says what to do if the application is rejected, but not clear on withdrawn. Unless this is being dealt with in another ticket, in which case ignore me. @MylesJarvis do you remember what we said we'd do with withdrawn application, can they add the same one again?

**Comment by MylesJarvis** [comment]
> Looks good! I just had that question about if we allow them to add the same application if they've withdrawn the same one already. At the moment you can't.
> 
> The ticket says what to do if the application is rejected, but not clear on withdrawn. Unless this is being dealt with in another ticket, in which case ignore me. @MylesJarvis do you remember what we said we'd do with withdrawn application, can they add the same one again?

Thanks Maeve! On withdrawn, we should ideally allow them to add the same choice again. I believe this led to a technical discussion on how we would treat application choices in this situation (convert the original application choice ID from withdrawn -> awaiting provider decision, if I remember correctly).

As James said, let's get this PR in and ticket up the add withdrawal logic, if we haven't already.

**Comment by JR-G** [comment]
> > Looks good! I just had that question about if we allow them to add the same application if they've withdrawn the same one already. At the moment you can't.
> > The ticket says what to do if the application is rejected, but not clear on withdrawn. Unless this is being dealt with in another ticket, in which case ignore me. @MylesJarvis do you remember what we said we'd do with withdrawn application, can they add the same one again?
> 
> Thanks Maeve! On withdrawn, we should ideally allow them to add the same choice again. I believe this led to a technical discussion on how we would treat application choices in this situation (convert the original application choice ID from withdrawn -> awaiting provider decision, if I remember correctly).
> 
> As James said, let's get this PR in and ticket up the add withdrawal logic, if we haven't already.

I've now added a card for this – happy to proceed with the PR

**Pull request merged**